src/Tools/ExternalAccess/Copilot/Internal/Analyzer/AbstractCopilotCodeAnalysisService.cs

```diff
@@ -144,7 +145,33 @@ public async Task<ImmutableArray<Diagnostic>> GetCachedDocumentDiagnosticsAsync(
             }
         }
 
-        return diagnostics.ToImmutable();
+        if (!span.HasValue)
+            return diagnostics.ToImmutable();
+
+        using var _2 = ArrayBuilder<Diagnostic>.GetInstance(out var filteredDiagnostics);
+        if (document.SupportsSyntaxTree)
+        {
+            // The location of Copilot diagnostics is on the method identifier, we'd like to expand the range to include them
+            // if any part of the method intersects with the given span.
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            foreach (var diagnostic in diagnostics)
+            {
+                var containingMethod = syntaxFacts.GetContainingMethodDeclaration(root, diagnostic.Location.SourceSpan.Start, useFullSpan: false);
+                if (containingMethod?.Span.IntersectsWith(span.Value) is true)
+                    filteredDiagnostics.Add(diagnostic);
+            }
+        }
+        else
+        {
+            foreach (var diagnostic in diagnostics)
+            {
+                if (diagnostic.Location.SourceSpan.IntersectsWith(span.Value))
+                    filteredDiagnostics.Add(diagnostic);
```

Comment:
fine for now.  but we really need this to be done through an ILangService so tha tthe *language* decides what span to be using instead.  right now this is really just not the right way to encode this sort of concept into a feature.

Reply:
Fixed